### PR TITLE
[Shardformer] Add parallel output for shardformer models(bloom, falcon)

### PR DESCRIPTION
**Pull request by Hz188**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [x] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`
Feature Devlopment, No related issue.

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

Add parallel cross entropy computation to bloom, falcon.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Review by ver217 — APPROVED** [review]

File: colossalai/shardformer/layer/loss.py [hunk 1]
```diff
@@ -86,7 +86,10 @@ def forward(
         dist.all_reduce(pred_logits, op=dist.ReduceOp.SUM, group=process_group)
         exp_logits = vocab_logits
         torch.exp(vocab_logits, out=exp_logits)
-        sum_exp_logits = torch.sum(exp_logits, dim=-1)
+        if exp_logits.dtype == torch.float16:
```
Comment (ver217) [inline colossalai/shardformer/layer/loss.py:89]: Can you add `dtype` to arguments, like https://pytorch.org/docs/stable/generated/torch.nn.functional.softmax.html#torch.nn.functional.softmax

Comment (Hz188) [reply colossalai/shardformer/layer/loss.py]: Finish it in [commit](https://github.com/hpcaitech/ColossalAI/pull/5702/commits/ca4b96ab6c48d9d87276a6256fc9b8f55e97aac0).
